coalib/parsing/FilterHelper.py

```diff
@@ -49,22 +50,83 @@ def apply_filter(filter_name, filter_args, all_bears=None):
+def _apply_section_filter(filter_name, filter_args, all_sections):
     """
-    Returns bears after filtering based on ``filters``. It returns
-    intersection of bears if more than one element is present in ``filters``
-    list.
+    Returns sections after filtering based on ``filter_args``. It
+    returns all sections if nothing is present in ``filter_args``.
+
+    :param filter_name:
+        Name of the section filter.
+    :param filter_args:
+        Arguments to be passed to the filter. For example:
+        ``['section_tags', ('save', 'change')]``
+    :param all_sections:
+        List of all sections on which filter is to be applied.
+    :return:
+        Filtered sections based on a single section filter.
+    """
+    if not is_valid_filter(filter_name):
+        raise InvalidFilterException(filter_name)
+    if not filter_args or len(filter_args) == 0:
+        return all_sections
+
+    filter_function = available_filters[filter_name]
+    filtered_sections = []
+
+    for section in all_sections:
+        if filter_function(section, filter_args):
+            filtered_sections += [section]
+
+    return filtered_sections
+
+
+def apply_filters(filters, bears=None, sections=None):
+    """
+    Returns bears or sections after filtering based on ``filters``.
+    It returns intersection if more than one element is present in
+    ``filters`` list. Either bears or sections need to be passed,
+    if both or none are passed it defaults to use bears gathering
+    and runs filter in bear filtering mode.
 
     :param filters:
-        List of args based on ``bears`` has to be filtered.
-        For example:
-        ``[['language', 'c', 'java'], ['can_fix', 'syntax']]``
+        List of args based on ``bears`` has to be filtered. For example:
+        ``[['language', 'c', 'java'], ['can_fix', 'syntax'],
+        ['section_tags', 'save']]``
     :param bears:
         The bears to filter.
+    :param sections:
+        The sections to filter.
     :return:
-        Filtered bears.
+        Filtered bears or sections.
     """
+    things = bears
+    applier = apply_filter
+    if sections is not None:
+        things = sections
+        applier = _apply_section_filter
+
     for filter in filters:
         filter_name, *filter_args = filter
-        bears = apply_filter(filter_name, filter_args, bears)
-    return bears
+        things = applier(filter_name, filter_args, things)
+    return things
+
+
+def collect_filters(args, arg_list=None, arg_parser=None):
+    """
+    Collects all filters from based on cli arguments.
+
+    :param args:
+        Parsed CLI args using which the filters are to be collected.
+    :param arg_list:
+        The CLI argument list.
+    :param arg_parser:
+        Instance of ArgParser that is used to parse arg list.
+    :return:
+        List of filters in standard filter format, i.e
+        ``[['filter_name', 'arg1', 'arg2']]``.
+    """
+    if args is None:
```

Comment:
Do we have any practical case for this condition ?
If so, is there way we can check this before calling this function ?

Reply:
```
run_coala

:param arg_parser:        Instance of ArgParser that is used to non-setting arguments.
:param arg_list:              The CLI argument list.
:param args:                   Alternative pre-parsed CLI arguments.
```

This is not just about the tests, since args to `run_coala()` can be optionally passed instead of arg_list, I think we need to merge opts from both, so as to enable the internal usage of this filter. Tests also use can use this by directly calling `run_coala()` with an arg list.